ddtrace/tracer/sampler_test.go

```diff
@@ -179,3 +184,426 @@ func TestRateSamplerSetting(t *testing.T) {
+					os.Setenv("DD_TRACE_SAMPLE_RATE", fmt.Sprint(rate))
+					defer os.Unsetenv("DD_TRACE_SAMPLE_RATE")
+					rs := newRulesSampler(rules)
+
+					span := makeSpan("http.request", "test-service")
+					result := rs.apply(span)
+					assert.True(result)
+					assert.Equal(rate, span.Metrics["_dd.rule_psr"])
+					if rate > 0.0 {
+						assert.Equal(0.5, span.Metrics["_dd.limit_psr"])
+					}
+				})
+			}
+		}
+	})
+}
+
+func TestRulesSamplerConcurrency(t *testing.T) {
+	rules := []SamplingRule{
+		ServiceRule("test-service", 1.0),
+		NameServiceRule("db.query", "postgres.db", 1.0),
+		NameRule("notweb.request", 1.0),
+	}
+	tracer := newTracer(WithSamplingRules(rules))
+	span := func(wg *sync.WaitGroup) {
+		defer wg.Done()
+		tracer.StartSpan("db.query", ServiceName("postgres.db")).Finish()
+	}
+
+	wg := &sync.WaitGroup{}
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go span(wg)
+	}
+	wg.Wait()
+}
+
+func TestRulesSamplerInternals(t *testing.T) {
+	makeSpanAt := func(op string, svc string, ts time.Time) *span {
+		s := newSpan(op, svc, "", 0, 0, 0)
+		s.Start = ts.UnixNano()
+		return s
+	}
+
+	t.Run("zero-rate", func(t *testing.T) {
+		assert := assert.New(t)
+		now := time.Now()
+		rs := &rulesSampler{}
+		span := makeSpanAt("http.request", "test-service", now)
+		rs.applyRate(span, 0.0, now)
+		assert.Equal(0.0, span.Metrics["_dd.rule_psr"])
+		_, ok := span.Metrics["_dd.limit_psr"]
+		assert.False(ok)
+	})
+
+	t.Run("full-rate", func(t *testing.T) {
+		assert := assert.New(t)
+		now := time.Now()
+		rs := newRulesSampler(nil)
+		// set samplingLimiter to specific state
+		rs.limiter.currentPeriod = now.Truncate(time.Second).Add(-1 * time.Second)
+		rs.limiter.previousRate = 1.0
+		rs.limiter.allowed = 1
+		rs.limiter.seen = 1
+
+		span := makeSpanAt("http.request", "test-service", now)
+		rs.applyRate(span, 1.0, now)
+		assert.Equal(1.0, span.Metrics["_dd.rule_psr"])
+		assert.Equal(1.0, span.Metrics["_dd.limit_psr"])
+	})
+
+	t.Run("limited-rate", func(t *testing.T) {
+		assert := assert.New(t)
+		now := time.Now()
+		rs := newRulesSampler(nil)
+		// force sampling limiter to 1.0 spans/sec
+		rs.limiter.limiter = rate.NewLimiter(rate.Limit(1.0), 1)
+		rs.limiter.currentPeriod = now.Truncate(time.Second).Add(-1 * time.Second)
+		rs.limiter.previousRate = 1.0
+		rs.limiter.allowed = 1
+		rs.limiter.seen = 1
+		// first span kept, second dropped
+		span := makeSpanAt("http.request", "test-service", now)
+		rs.applyRate(span, 1.0, now)
+		assert.EqualValues(ext.PriorityAutoKeep, span.Metrics[keySamplingPriority])
+		assert.Equal(1.0, span.Metrics["_dd.rule_psr"])
+		assert.Equal(1.0, span.Metrics["_dd.limit_psr"])
+		span = makeSpanAt("http.request", "test-service", now)
+		rs.applyRate(span, 1.0, now)
+		assert.EqualValues(ext.PriorityAutoReject, span.Metrics[keySamplingPriority])
+		assert.Equal(1.0, span.Metrics["_dd.rule_psr"])
+		assert.Equal(0.75, span.Metrics["_dd.limit_psr"])
+	})
+}
+
+func TestSamplingLimiter(t *testing.T) {
```

Comment:
Please add a wide range of table tests which cover many scenarios:
* On start
* After start
* With previous rate
* Without previous rate
And let's make sure that we cover everything and it works correctly.